### DISCUSSION TITLE
Can I reproduce the same results from the webui with the same parameters?

### DISCUSSION THREAD

**Discussion by zgxkbtl** [comfyanonymous/ComfyUI#118] [Q&A] [answered]
Can I reproduce the same results from the webui with the same parameters?
I have previously used AUTOMATIC1111/stable-diffusion-webui and recently tried ComfyUI, which seems to have powerful features. However, when I use the same parameters as in the webui (model, prompt, random seed, canvas size, sampling method, etc.), I still can't reproduce the results from the webui.

By the way, does the "clip skip" parameter exist in ComfyUI?

**Comment by comfyanonymous** [comment, marked as answer]
>Can I reproduce the same results from the webui with the same parameters?

No. Noise on ComfyUI is generated on the CPU while the a1111 UI generates it on the GPU. This makes ComfyUI seeds reproducible across different hardware configurations but makes them different from the ones used by the a1111 UI.

I didn't care about having compatibility with the a1111 UI seeds because that UI has broken seeds quite a few times now so it seemed like a hassle to do so.


>By the way, does the "clip skip" parameter exist in ComfyUI?

Yes, it's the CLIPSetLastLayer node.

**Reply by zgxkbtl** [reply]
Thank you for your answer. 
I'd like to ask if syntax like [wedding dress: china dress: 0.5] can be correctly recognized in ComfyUI? Which is an element fusion syntax , I tried it, and the result doesn't seem to be particularly good?

**Reply by comfyanonymous** [reply]
No the only prompt syntax from a1111 that I have are the (weights:1.3) of the words in the prompt. and the {wildcard|prompt} syntax

If you want to change prompts/models/Loras/samplers/etc... during sampling you can use the KSamplerAdvanced node for that.

**Reply by catboxanon** [reply]
Would you consider adding the option to generate the noise on the GPU? Or do you not want to expose such an option to everyone? I just did a quick test modifying the randn call and the result is identical to A1111. I'd understand if you don't want to expose that though incase many users were to start utilizing it.

**Reply by comfyanonymous** [reply]
It's not something I want to support in master because it doesn't make sense to generate noise on the GPU.

**Reply by Lling00** [reply]
> Would you consider adding the option to generate the noise on the GPU? Or do you not want to expose such an option to everyone? I just did a quick test modifying the randn call and the result is identical to A1111. I'd understand if you don't want to expose that though incase many users were to start utilizing it

How did you do it? I also want to achieve the same effect as A1111, because when I communicate with others, they use A1111 and the results are different, which makes communication more difficult.

**Reply by catboxanon** [reply]
@Lling00
Change this line here:
https://github.com/comfyanonymous/ComfyUI/blob/8cb54fa2faadbb1b87a3c64e6857598887bbcecb/nodes.py#L688

```diff
- noise = torch.randn(latent_image.size(), dtype=latent_image.dtype, layout=latent_image.layout, generator=torch.manual_seed(seed), device="cpu")
+ noise = torch.randn(latent_image.size(), dtype=latent_image.dtype, layout=latent_image.layout, generator=torch.cuda.manual_seed(seed), device=device)
```

If you use weighting in your prompt, you'll need to use the custom node mentioned here: https://github.com/comfyanonymous/ComfyUI/discussions/521

**Reply by Lling00** [reply]
Thank you

**Reply by sdqq1234** [reply]
I changed this line in the new version, when only one image is generated, it can be the same as the webUI result, but when multiple images are generated, only the first one is the same, and the others are different. why?  
new version is here at [ line 12](https://github.com/comfyanonymous/ComfyUI/blob/master/comfy/sample.py#L12)
Where does WebUI calculate noise?@catboxanon

**Comment by catboxanon** [comment]
Update on this: with the addition of the noise nodes available [here](https://github.com/BlenderNeko/ComfyUI_Noise) and weighting nodes available [here](https://github.com/BlenderNeko/ComfyUI_ADV_CLIP_emb), the web UI now also supports the [choice between using GPU and CPU for noise](https://github.com/AUTOMATIC1111/stable-diffusion-webui/blob/5ab7f213bec2f816f9c5644becb32eb72c8ffb89/modules/shared.py#L354), so results can be replicated both ways.

**Reply by EXOgreen** [reply]
Any chance you could explain in more detail how to achieve this? I've been messing with the nodes a bit, but I don't know where to find the right values from the automatic1111 ui.

**Reply by cmelle** [reply]
in automatic1111, go on settings > Stable Diffusion, and on the bottom you'll see "Random number generator source. Changes seeds drastically. Use CPU to produce the same picture across different vidocard vendors."
Then I managed to produce the same prompt between ComfyUI and automatic1111 webUI, with sampler dpmpp_2m karras, but with a very simple setting.
Dont' forget that there is no "eta noise seed delta" in comfyUI
For example for now I didn't manage to have the same results when I use embeddings.

**Reply by MattRM2** [reply]
For me it's also something like 70-80%. I don't know why but the quality of shading / lighting is not has good as A1111. But I can run SDXL in ComfyUI without problem and have an out of memory error in A1111. But my goal is the quality of shading / lighting.

**Reply by MattRM2** [reply]
Ok, I found why I had bad result from ComfyUI... My negative prompt have in A1111 weight that f.cked my image in ComfyUI. Remove all my weight, I have at 95% the result I have in A1111. Just a problem of weight in prompt. I never think about this and how much it can impact my result. So I will make custom negetive prompt that avoid weights inside. I'm happy because I prefer ComfyUI for his workflow and the speed.

Here the result now : 
![Capture d’écran 2023-08-06 143053](https://github.com/comfyanonymous/ComfyUI/assets/25594057/1e7623b7-e03d-43d5-ab07-cd6ec6665815)

**Reply by barryd15** [reply]
Thanks for posting this.  I want to improve support for pasting txt2img generation parameters / PNGs into ComfyUI, and this will help.

**Reply by ousinkou** [reply]
I also get 95% the result I have in A1111. I wonder how the left 5% difference happen.

**Comment by MattRM2** [comment]
Hi,

For the clip skip in A1111 set at 1, how to setup the same in ComfyUI using CLIPSetLastLayer ? Does the clip skip 1 in A1111 is -1 in ComfyUI? Could you give me some more info to setup it at the same ?

Thx

Matt

**Reply by comfyanonymous** [reply]
Yes, clip skip 1 means CLIPSetLastLayer -1, clip skip 2 means CLIPSetLastLayer -2 and so on.

**Reply by MattRM2** [reply]
Oki, thank for your answer Comfyanonymous.

**Comment by Shariq81** [comment]
![image](https://github.com/comfyanonymous/ComfyUI/assets/31780478/1ba9e291-bb9b-4398-9fe6-03dc9aa7c5c8)
The first one is from ComfyUI and the second is from Automatic1111. All other settings same.

**Reply by ltdrdata** [reply]
FYI, simply copying and pasting the prompt and using the same CFG and sampler does not guarantee the same settings. There are bunch of different things.

**Comment by buggsi** [comment]
Efficient Loader has that now https://github.com/jags111/efficiency-nodes-comfyui you can install it from the Manager.

![image](https://github.com/user-attachments/assets/92490580-054f-4c61-a8e8-112fc50d41f5)

**Comment by nailz420** [comment]
As far as I know, different torch versions produce different results also
